Reader: Add UI for the Reader tags feed

wargcm opened the pull request:
Part of #23013

## Description

Adds the UI for the "Tags" feed.

Future PRs:
- Card actions
- "More" end cell
- Accessibility
- Empty feed state
- Fetching posts for tags
- Loading placeholders
- Consolidating display functions between `ReaderPostCardCell` & `ReaderTagCell`
- Analytics

## Video

<details><summary>iPhone</summary>

https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/b930721a-2731-444e-aab0-c9279180383e
</details>

<details><summary>iPad</summary>

https://github.com/wordpress-mobile/WordPress-iOS/assets/2454408/3c3ecade-3f95-48fc-ab49-3bfd5de17280
</details>


## Testing

**Test setup:**
You'll need some cached tag posts to see any UI:
- Navigate to `Discover`
- Tap on a post
- Tap on one of the tags for that post
- If you aren't following it, follow the tag

**To test:**
- Launch Jetpack and login
- Complete the **Test setup**
- Navigate to the "Your Tags" Reader feed
- Scroll to the tag which has posts saved locally
- 🔎 **Verify** the UI matches the designs for the new cell
- Find a post with no featured image
- 🔎 **Verify** the text is expanded
- Find a post where the content doesn't fit the full height
- 🔎 **Verify**:
  - The number of likes/comments label is aligned to the bottom
  - The like button and menu button are aligned to the bottom

## Regression Notes
1. Potential unintended areas of impact
Should be none.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A

3. What automated tests I added (or what prevented me from doing so)
N/A
PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

wargcm committed:
Add UI for the Reader tags feed

wargcm committed:
Recreate results controller if `entityName` changes

When a table view handler supports more than one type of entity, the
results controller needs to be recreated when that entity changes. Otherwise,
it can cause a crash when fetching the new type of entity.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dvdchr commented on WordPress/Classes/ViewRelated/Reader/ReaderTagCardCell.swift:
```diff
@@ -0,0 +1,125 @@
+        collectionView.register(nib, forCellWithReuseIdentifier: Constants.cellIdentifier)
+    }
+
+    func setupDataSource() {
```
Nit: To prevent force unwrapping on `dataSource`, what do you think about moving this to the `dataSource`'s lazy initialization body instead? The `dataSource` will be called via `controller(_:didChangeContentWith:)`, so it will be initialized.

wargcm replied:
Good call! I've updated that in: https://github.com/wordpress-mobile/WordPress-iOS/pull/23014/commits/8212343d84b036d7609dcf95f288f593bb329b10

dvdchr reviewed (approved):
Overall the code works as described, and it looks good to me. I left some minor comments and also noticed this one small design detail where there should've been a wider gap. However, this can be addressed later since it's non-breaking and still a work in progress.

Debug | Design
-|-
<img width="280" alt="Screenshot 2024-04-22 at 15 50 22" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/c35d0925-86a2-45db-b91b-50893b9a2e33"> | <img width="318" alt="Screenshot 2024-04-22 at 15 03 32" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/321abfeb-a13b-490a-94a2-3ad544db398d">

wargcm committed:
Update data source to be a lazy variable

wargcm committed:
Update spacing for `ReaderTagCardCell`

wargcm commented:
Thanks, @dvdchr! I fixed the spacing in https://github.com/wordpress-mobile/WordPress-iOS/pull/23014/commits/0311bf329eeedcbfff7032337e423d94fdf4c959. I realized I missed another thing when checking the designs again. In the designs, the content expands to the edges of the screen when scrolling the collection view. So I've also removed the clipping on the collection view to match that behavior.
